File: src/exdl_alut.erl
%%
%%  Copyright (c) 2008 David Koch
%%
%%  See the file "license.terms" for information on usage and redistribution
%%  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
%% 
%%     $Id: exdl_alut.erl,v 1.2 2008/11/06 07:49:22 kochise Exp $
%%

-module(exdl_alut).

-include("exdl.hrl").

-ifdef(COMPILE_ALUT).

%% 1- EXPORT

%-compile(export_all).

-export( [ init/2
         , initWithoutContext/2
         , exit/0
         , getError/0
         , getErrorString/1
         , createBufferFromFile/1
         , createBufferFromFileImage/2
         , createBufferHelloWorld/0
         , createBufferWaveform/4
         , loadMemoryFromFile/4
         , loadMemoryFromFileImage/5
         , loadMemoryHelloWorld/3
         , loadMemoryWaveform/7
         , getMIMETypes/1
         , getMajorVersion/0
         , getMinorVersion/0
         , sleep/1
         ]
       ).

%% 2- INCLUDE

-include("exdl_al.hrl").
-include("exdl_util.hrl").

-include("exdl_alut.hrl").

%% 3- IMPORT

-import( exdl_driver
       , [ call/2
         , cast/2
         ]
       ).

-import( exdl_util
       , [ bin2list/2
         , bin2list/3
         , term2bin/2
         , term2bin/3
         , matrix2bin/2
         ]
       ).

%% 4- FUNCTIONS

%% Func:       init
%% Args:       Argcp, Argv
%% Returns:    Exdl
%% C-API func: ALboolean alutInit(int* argcp, char** argv)

init(Argcp, Argv)
 -> case
      call
      ( ?alutInit
      , <<Argcp:?_PTR
        , Argv:?_PTR
        >>
      )
    of
      <<Exdl:?ALboolean
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       initWithoutContext
%% Args:       Argcp, Argv
%% Returns:    Exdl
%% C-API func: ALboolean alutInitWithoutContext(int* argcp, char** argv)

initWithoutContext(Argcp, Argv)
 -> case
      call
      ( ?alutInitWithoutContext
      , <<Argcp:?_PTR
        , Argv:?_PTR
        >>
      )
    of
      <<Exdl:?ALboolean
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       exit
%% Args:       none
%% Returns:    Exdl
%% C-API func: ALboolean alutExit(void)

exit()
 -> case
      call
      ( ?alutExit
      , [
        ]
      )
    of
      <<Exdl:?ALboolean
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getError
%% Args:       none
%% Returns:    Exdl
%% C-API func: ALenum alutGetError(void)

getError()
 -> case
      call
      ( ?alutGetError
      , [
        ]
      )
    of
      <<Exdl:?ALenum
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getErrorString
%% Args:       Error
%% Returns:    Exdl
%% C-API func: const char* alutGetErrorString(ALenum error)

getErrorString(Error)
 -> case
      call
      ( ?alutGetErrorString
      , <<Error:?ALenum
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?String
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       createBufferFromFile
%% Args:       FileName
%% Returns:    Exdl
%% C-API func: ALuint alutCreateBufferFromFile(const char* fileName)

createBufferFromFile(FileName)
 -> case
      call
      ( ?alutCreateBufferFromFile
      , <<FileName:?_PTR
        >>
      )
    of
      <<Exdl:?ALuint
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       createBufferFromFileImage
%% Args:       Data, Length
%% Returns:    Exdl
%% C-API func: ALuint alutCreateBufferFromFileImage(const ALvoid* data, ALsizei length)

createBufferFromFileImage(Data, Length)
 -> case
      call
      ( ?alutCreateBufferFromFileImage
      , <<Length:?ALsizei
        , Data:?_PTR
        >>
      )
    of
      <<Exdl:?ALuint
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       createBufferHelloWorld
%% Args:       none
%% Returns:    Exdl
%% C-API func: ALuint alutCreateBufferHelloWorld(void)

createBufferHelloWorld()
 -> case
      call
      ( ?alutCreateBufferHelloWorld
      , [
        ]
      )
    of
      <<Exdl:?ALuint
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       createBufferWaveform
%% Args:       Waveshape, Frequency, Phase, Duration
%% Returns:    Exdl
%% C-API func: ALuint alutCreateBufferWaveform(ALenum waveshape, ALfloat frequency, ALfloat phase, ALfloat duration)

createBufferWaveform(Waveshape, Frequency, Phase, Duration)
 -> case
      call
      ( ?alutCreateBufferWaveform
      , <<Waveshape:?ALenum
        , Frequency:?ALfloat
        , Phase:?ALfloat
        , Duration:?ALfloat
        >>
      )
    of
      <<Exdl:?ALuint
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       loadMemoryFromFile
%% Args:       FileName, Format, Size, Frequency
%% Returns:    Exdl
%% C-API func: ALvoid* alutLoadMemoryFromFile(const char* fileName, ALenum* format, ALsizei* size, ALfloat* frequency)

loadMemoryFromFile(FileName, Format, Size, Frequency)
 -> case
      call
      ( ?alutLoadMemoryFromFile
      , <<FileName:?_PTR
        , Format:?_PTR
        , Size:?_PTR
        , Frequency:?_PTR
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       loadMemoryFromFileImage
%% Args:       Data, Length, Format, Size, Frequency
%% Returns:    Exdl
%% C-API func: ALvoid* alutLoadMemoryFromFileImage(const ALvoid* data, ALsizei length, ALenum* format, ALsizei* size, ALfloat* frequency)

loadMemoryFromFileImage(Data, Length, Format, Size, Frequency)
 -> case
      call
      ( ?alutLoadMemoryFromFileImage
      , <<Data:?_PTR
        , Length:?ALsizei
        , Format:?_PTR
        , Size:?_PTR
        , Frequency:?_PTR
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       loadMemoryHelloWorld
%% Args:       Format, Size, Frequency
%% Returns:    Exdl
%% C-API func: ALvoid* alutLoadMemoryHelloWorld(ALenum* format, ALsizei* size, ALfloat* frequency)

loadMemoryHelloWorld(Format, Size, Frequency)
 -> case
      call
      ( ?alutLoadMemoryHelloWorld
      , <<Format:?_PTR
        , Size:?_PTR
        , Frequency:?_PTR
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       loadMemoryWaveform
%% Args:       Waveshape, Frequency, Phase, Duration, Format, Size, Freq
%% Returns:    Exdl
%% C-API func: ALvoid* alutLoadMemoryWaveform(ALenum waveshape, ALfloat frequency, ALfloat phase, ALfloat duration, ALenum* format, ALsizei* size, ALfloat* freq)

loadMemoryWaveform(Waveshape, Frequency, Phase, Duration, Format, Size, Freq)
 -> case
      call
      ( ?alutLoadMemoryWaveform
      , <<Waveshape:?ALenum
        , Frequency:?ALfloat
        , Phase:?ALfloat
        , Duration:?ALfloat
        , Format:?_PTR
        , Size:?_PTR
        , Freq:?_PTR
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getMIMETypes
%% Args:       Loader
%% Returns:    Exdl
%% C-API func: const char* alutGetMIMETypes(ALenum loader)

getMIMETypes(Loader)
 -> case
      call
      ( ?alutGetMIMETypes
      , <<Loader:?ALenum
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?String
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getMajorVersion
%% Args:       none
%% Returns:    Exdl
%% C-API func: ALint alutGetMajorVersion(void)

getMajorVersion()
 -> case
      call
      ( ?alutGetMajorVersion
      , [
        ]
      )
    of
      <<Exdl:?ALint
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getMinorVersion
%% Args:       none
%% Returns:    Exdl
%% C-API func: ALint alutGetMinorVersion(void)

getMinorVersion()
 -> case
      call
      ( ?alutGetMinorVersion
      , [
        ]
      )
    of
      <<Exdl:?ALint
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       sleep
%% Args:       Duration
%% Returns:    Exdl
%% C-API func: ALboolean alutSleep(ALfloat duration)

sleep(Duration)
 -> case
      call
      ( ?alutSleep
      , <<Duration:?ALfloat
        >>
      )
    of
      <<Exdl:?ALboolean
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

-endif. % COMPILE_ALUT

File: priv/src/exdl_alut.c
/*
 *  Copyright (c) 2008 David Koch
 *
 *  See the file "license.terms" for information on usage and redistribution
 *  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 *     $Id: exdl_alut.c,v 1.1 2008/11/06 07:49:22 kochise Exp $
 */

#include "exdl.h"

#ifdef COMPILE_ALUT

EXDL_API(ealut_init)
{
  /* ALboolean alutInit(int* argcp, char** argv) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(int,   argcp)
    STACK_FMT_DATA(char*, argv[])
  STACK_FMT_END()

  CATCH_RESULT(ALboolean)

    alutInit
    ( STACK_GET_ADDR(argcp)
    , STACK_GET_ADDR(argv)
    );

  THROW_RESULT(ALboolean)
}

EXDL_API(ealut_initWithoutContext)
{
  /* ALboolean alutInitWithoutContext(int* argcp, char** argv) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(int,   argcp)
    STACK_FMT_DATA(char*, argv[])
  STACK_FMT_END()

  CATCH_RESULT(ALboolean)

    alutInitWithoutContext
    ( STACK_GET_ADDR(argcp)
    , STACK_GET_ADDR(argv)
    );

  THROW_RESULT(ALboolean)
}

EXDL_API(ealut_exit)
{
  /* ALboolean alutExit(void) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  CATCH_RESULT(ALboolean)

    alutExit
    (
    );

  THROW_RESULT(ALboolean)
}

EXDL_API(ealut_getError)
{
  /* ALenum alutGetError(void) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  CATCH_RESULT(ALenum)

    alutGetError
    (
    );

  THROW_RESULT(ALenum)
}

EXDL_API(ealut_getErrorString)
{
  /* const char* alutGetErrorString(ALenum error) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALenum, error)
  STACK_FMT_END()

  CATCH_RESULT(const char*)

    alutGetErrorString
    ( STACK_GET_DATA(error)
    );

  THROW_STRING()
}

EXDL_API(ealut_createBufferFromFile)
{
  /* ALuint alutCreateBufferFromFile(const char* fileName) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(char, fileName[])
  STACK_FMT_END()

  CATCH_RESULT(ALuint)

    alutCreateBufferFromFile
    ( STACK_GET_ADDR(fileName)
    );

  THROW_RESULT(ALuint)
}

EXDL_API(ealut_createBufferFromFileImage)
{
  /* ALuint alutCreateBufferFromFileImage(const ALvoid* data, ALsizei length) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALsizei, length)
    STACK_FMT_DATA(ALvoid,  data[])
  STACK_FMT_END()

  CATCH_RESULT(ALuint)

    alutCreateBufferFromFileImage
    ( (sd->next_bin == 0)
    ? STACK_GET_ADDR(data)
    : (ALvoid*) sd->bin[0].base
    , STACK_GET_DATA(length)
    );

  THROW_RESULT(ALuint)
}

EXDL_API(ealut_createBufferHelloWorld)
{
  /* ALuint alutCreateBufferHelloWorld(void) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  CATCH_RESULT(ALuint)

    alutCreateBufferHelloWorld
    (
    );

  THROW_RESULT(ALuint)
}

EXDL_API(ealut_createBufferWaveform)
{
  /* ALuint alutCreateBufferWaveform(ALenum waveshape, ALfloat frequency, ALfloat phase, ALfloat duration) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALenum,  waveshape)
    STACK_FMT_DATA(ALfloat, frequency)
    STACK_FMT_DATA(ALfloat, phase)
    STACK_FMT_DATA(ALfloat, duration)
  STACK_FMT_END()

  CATCH_RESULT(ALuint)

    alutCreateBufferWaveform
    ( STACK_GET_DATA(waveshape)
    , STACK_GET_DATA(frequency)
    , STACK_GET_DATA(phase)
    , STACK_GET_DATA(duration)
    );

  THROW_RESULT(ALuint)
}

EXDL_API(ealut_loadMemoryFromFile)
{
  /* ALvoid* alutLoadMemoryFromFile(const char* fileName, ALenum* format, ALsizei* size, ALfloat* frequency) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALenum*,  format)
    STACK_FMT_DATA(ALsizei*, size)
    STACK_FMT_DATA(ALfloat*, frequency)
    STACK_FMT_DATA(char,     fileName[])
  STACK_FMT_END()

  CATCH_RESULT(ALvoid*)

    alutLoadMemoryFromFile
    ( STACK_GET_ADDR(fileName)
    , STACK_GET_DATA(format)
    , STACK_GET_DATA(size)
    , STACK_GET_DATA(frequency)
    );

  THROW_POINTER()
}

EXDL_API(ealut_loadMemoryFromFileImage)
{
  /* ALvoid* alutLoadMemoryFromFileImage(const ALvoid* data, ALsizei length, ALenum* format, ALsizei* size, ALfloat* frequency) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALsizei,  length)
    STACK_FMT_DATA(ALenum*,  format)
    STACK_FMT_DATA(ALsizei*, size)
    STACK_FMT_DATA(ALfloat*, frequency)
    STACK_FMT_DATA(ALvoid,   data[])
  STACK_FMT_END()

  CATCH_RESULT(ALvoid*)

    alutLoadMemoryFromFileImage
    ( (sd->next_bin == 0)
    ? STACK_GET_ADDR(data)
    : (ALvoid*) sd->bin[0].base
    , STACK_GET_DATA(length)
    , STACK_GET_DATA(format)
    , STACK_GET_DATA(size)
    , STACK_GET_DATA(frequency)
    );

  THROW_POINTER()
}

EXDL_API(ealut_loadMemoryHelloWorld)
{
  /* ALvoid* alutLoadMemoryHelloWorld(ALenum* format, ALsizei* size, ALfloat* frequency) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALenum*,  format)
    STACK_FMT_DATA(ALsizei*, size)
    STACK_FMT_DATA(ALfloat*, frequency)
  STACK_FMT_END()

  CATCH_RESULT(ALvoid*)

    alutLoadMemoryHelloWorld
    ( STACK_GET_DATA(format)
    , STACK_GET_DATA(size)
    , STACK_GET_DATA(frequency)
    );

  THROW_POINTER()
}

EXDL_API(ealut_loadMemoryWaveform)
{
  /* ALvoid* alutLoadMemoryWaveform(ALenum waveshape, ALfloat frequency, ALfloat phase, ALfloat duration, ALenum* format, ALsizei* size, ALfloat* freq) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALenum,   waveshape)
    STACK_FMT_DATA(ALfloat,  frequency)
    STACK_FMT_DATA(ALfloat,  phase)
    STACK_FMT_DATA(ALfloat,  duration)
    STACK_FMT_DATA(ALenum*,  format)
    STACK_FMT_DATA(ALsizei*, size)
    STACK_FMT_DATA(ALfloat*, freq)
  STACK_FMT_END()

  CATCH_RESULT(ALvoid*)

    alutLoadMemoryWaveform
    ( STACK_GET_DATA(waveshape)
    , STACK_GET_DATA(frequency)
    , STACK_GET_DATA(phase)
    , STACK_GET_DATA(duration)
    , STACK_GET_DATA(format)
    , STACK_GET_DATA(size)
    , STACK_GET_DATA(freq)
    );

  THROW_POINTER()
}

EXDL_API(ealut_getMIMETypes)
{
  /* const char* alutGetMIMETypes(ALenum loader) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALenum, loader)
  STACK_FMT_END()

  CATCH_RESULT(const char*)

    alutGetMIMETypes
    ( STACK_GET_DATA(loader)
    );

  THROW_STRING()
}

EXDL_API(ealut_getMajorVersion)
{
  /* ALint alutGetMajorVersion(void) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  CATCH_RESULT(ALint)

    alutGetMajorVersion
    (
    );

  THROW_RESULT(ALint)
}

EXDL_API(ealut_getMinorVersion)
{
  /* ALint alutGetMinorVersion(void) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  CATCH_RESULT(ALint)

    alutGetMinorVersion
    (
    );

  THROW_RESULT(ALint)
}

EXDL_API(ealut_sleep)
{
  /* ALboolean alutSleep(ALfloat duration) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(ALfloat, duration)
  STACK_FMT_END()

  CATCH_RESULT(ALboolean)

    alutSleep
    ( STACK_GET_DATA(duration)
    );

  THROW_RESULT(ALboolean)
}

#endif // COMPILE_ALUT
